### DISCUSSION TITLE
restrict the percentage of patient for specific attribute

### DISCUSSION THREAD

**Discussion by idhamari** [synthetichealth/synthea#1302] [Q&A]
I have two questions regarding designing a module for a specific disease.

- How can I define a state that restricts the patient's attribute such as gender or race to a specific probability e.g.  gender1: 0.99 and gender2: 0.01. 
- How can I set probabilities for death based on age groups e.g. age-group1:0.1.agegroup2: 0.2,..etc.

**Comment by jawalonoski** [comment]
The modules don't define demographics of patients... the demographics are determined by census data.

Keep in mind that the following example is not a complete working module... as-is, it does nothing.

If you want a module to transition to a state based on existing demographics, you could do something like this:

<img width="401" alt="image" src="https://github.com/synthetichealth/synthea/assets/1054765/1a719dfa-c908-40d5-a30f-3fe2879a77c5">

```
{
  "name": "Untitled",
  "remarks": [
    "A blank module"
  ],
  "states": {
    "Initial": {
      "type": "Initial",
      "complex_transition": [
        {
          "condition": {
            "condition_type": "Gender",
            "gender": "F"
          },
          "distributions": [
            {
              "transition": "Yes",
              "distribution": 0.99
            },
            {
              "transition": "No",
              "distribution": 0.01
            }
          ]
        },
        {
          "condition": {
            "condition_type": "Gender",
            "gender": "M"
          },
          "distributions": [
            {
              "transition": "Yes",
              "distribution": 0.01
            },
            {
              "transition": "No",
              "distribution": 0.99
            }
          ]
        }
      ]
    },
    "Terminal": {
      "type": "Terminal"
    },
    "Yes": {
      "type": "Simple",
      "direct_transition": "Terminal"
    },
    "No": {
      "type": "Simple",
      "direct_transition": "Terminal"
    }
  },
  "gmf_version": 2
}
```

**Comment by jawalonoski** [comment]
If you want a module to define probability of death based on age, you could do something like the below example.

Keep in mind that the following example is not a complete working module... as-is, it does not work.

<img width="573" alt="image" src="https://github.com/synthetichealth/synthea/assets/1054765/e891822c-c48c-4513-a701-cea40fd311bf">

```
{
  "name": "Untitled",
  "remarks": [
    "A blank module"
  ],
  "states": {
    "Initial": {
      "type": "Initial",
      "direct_transition": "Wait"
    },
    "Terminal": {
      "type": "Terminal"
    },
    "Death": {
      "type": "Death",
      "direct_transition": "Terminal"
    },
    "Wait": {
      "type": "Guard",
      "allow": {
        "condition_type": "Attribute",
        "attribute": "some_variable_set_elsewhere",
        "operator": "==",
        "value": true
      },
      "complex_transition": [
        {
          "condition": {
            "condition_type": "Age",
            "operator": "<=",
            "quantity": 10,
            "unit": "years",
            "value": 0
          },
          "distributions": [
            {
              "transition": "Terminal",
              "distribution": 0.9
            },
            {
              "transition": "Death",
              "distribution": 0.1
            }
          ]
        },
        {
          "condition": {
            "condition_type": "Age",
            "operator": "<=",
            "quantity": 20,
            "unit": "years",
            "value": 0
          },
          "distributions": [
            {
              "transition": "Terminal",
              "distribution": 0.8
            },
            {
              "transition": "Death",
              "distribution": 0.2
            }
          ]
        }
      ]
    }
  },
  "gmf_version": 2
}
```

**Comment by idhamari** [comment]
Thanks for your quick reply. I will try that!

**Comment by idhamari** [comment]
Based on the information above and following some examples, I created this simple module. The patient must have this disease based on the sick state (the probabilities are the prevalence) :

<img width="268" alt="mymodule" src="https://github.com/synthetichealth/synthea/assets/28802516/238f55ef-a785-42ef-b20f-bc8c827a1ec4">

I have a few questions: 
1. How can I generate a dataset that has only the patients with this disease and their information e.g. status, p1 and p2.? 
   - when I run using: 
   
             run_synthea  -k ./mymodule.json Alabama
   
   I got this error: 

         Waiting for threads to finish... java.util.concurrent.ThreadPoolExecutor@ee8e7ff[Shutting down, pool size = 1, active threads = 1,  queued tasks = 0, completed tasks = 0]
         Waiting for threads to finish... java.util.concurrent.ThreadPoolExecutor@ee8e7ff[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 0]
         Waiting for threads to finish... java.util.concurrent.ThreadPoolExecutor@ee8e7ff[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 0]
         Waiting for threads to finish... java.util.concurrent.ThreadPoolExecutor@ee8e7ff[Shutting down, pool size = 1, active threads = 1, queued tasks = 0, completed tasks = 0]
         java.lang.RuntimeException: Failed to produce a matching patient after 1000 attempts. Ensure that it is possible for all requested   demographics to meet the criteria. (e.g., make sure there is no age restriction that conflicts with a requested condition, such as limiting age  to 0-18 and requiring all patients have a condition that only onsets after 55.) If you are confident that the constraints are possible to  satisfy but rare, consider increasing the value in config setting `generate.max_attempts_to_keep_patient`
        at org.mitre.synthea.engine.Generator.generatePerson(Generator.java:502)
        at org.mitre.synthea.engine.Generator.lambda$run$3(Generator.java:382)
        at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
        at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
        at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
        at java.base/java.lang.Thread.run(Thread.java:833)
        Records: total=0, alive=0, dead=0
        RNG=1
        Clinician RNG=6750

3. Will the patient's distributions e.g. gender and race follow the probabilities in the states e.g. in Race and Sex states?
4. In case one didn't use the population argument, will the number of patients follows the prevalence of the disease?

**Comment by jawalonoski** [comment]
First, by using the `-k` flag, you are using your module as a _Keep Module_ instead of a _Disease Module_.

The module you have written looks like it is designed to defined disease prevalence and morbidity, so you should run it as a normal module. See https://github.com/synthetichealth/synthea/wiki/Generic-Module-Framework%3A-Running-New-Modules

If you meant to define a keep module, you should carefully reread this page: https://github.com/synthetichealth/synthea/wiki/Keep-Patients-Module. For starters, you don't define a "Keep" state, so you'll never generate an acceptable patient (i.e., 
`java.lang.RuntimeException: Failed to produce a matching patient after 1000 attempts. Ensure that it is possible for all requested   demographics to meet the criteria.`

Otherwise, the module you have presented probably won't work as you intend, because you never specify when/where the Condition is actually diagnosed (i.e., written into the health record). There are many ways to do this, but you need an `Encounter`. In the example below, I specify that the diagnosis is made at the next wellness encounter.

![image](https://github.com/synthetichealth/synthea/assets/1054765/9f4e45cd-b797-4ab8-a633-b2f96bf4d5c2)
